doc/langref.html.in

```diff
@@ -6899,8 +6899,255 @@ fn foo() i32 {
 
       {#header_open|Result Location Semantics#}
       <p>
-      <a href="https://github.com/ziglang/zig/issues/2809">TODO add documentation for this</a>
+      During compilation, every Zig expression and sub-expression is assigned optional result location
+      information. This information dictates what type the expression should have (its result type), and
+      where the resulting value should be placed in memory (its result location). This is not an
+      implementation detail: this logic is codified into the Zig language specification. This system is
+      collectively referred to as "Result Location Semantics".
       </p>
+      {#header_open|Result Types#}
+      <p>
+      Result types are propagated recursively through expressions where possible. For instance, if the
+      expression {#syntax#}&e{#endsyntax#} has result type {#syntax#}*u32{#endsyntax#}, then
+      {#syntax#}e{#endsyntax#} is given a result type of {#syntax#}u32{#endsyntax#}, allowing the
+      language to perform this coercion before taking a reference.
+      </p>
+      <p>
+      The result type mechanism is utilized by casting builtins such as {#syntax#}@intCast{#endsyntax#}.
+      Rather than taking as an argument the type to cast to, these builtins use their result type to
+      determine this information. The result type is often known from context; where it is not, the
+      {#syntax#}@as{#endsyntax#} builtin can be used to explicitly provide a result type.
+      </p>
+      <p>
+      We can break down the result types for each component of a simple expression as follows:
+      </p>
+      {#code_begin|test|result_type_propagation#}
+test "result type propagates through struct initializer" {
+    const S = struct { x: u32 };
+    const val: u64 = 123;
+    const s: S = .{ .x = @intCast(val) };
+    // .{ .x = @intCast(val) }   has result type `S` due to the type annotation
+    //         @intCast(val)     has result type `u32` due to the type of the field `S.x`
+    //                  val      has no result type, as it is permitted to be any integer type
```

Comment:
Can you explain how the `u64` annotation on `val` factors in here?  I'm reading the existing text as saying that the `@intCast()` only expects an "integer", so it doesn't really constrain the `val`.  But doesn't the `u64` play a part?  (Also, why isn't the "must be an integer" represented in RLS?  Having "no result type" implies its an anytype?)

Reply:
The `u64` annotation on `val` isn't related to the line of code in question here, it's literally just... making `val` a `u64`. Is this somehow unclear?

> Also, why isn't the "must be an integer" represented in RLS?

RLS, in general, can only give *concrete types* as result types. "Some kind of integer" is not a specific type in Zig. Even if RLS could represent this, it's not actually a particularly helpful concept - I can't think of anywhere this result type would have an effect.

Reply:
Ah, this is why I was confused about "result types" being different from regular "Zig types".  I assumed `@intCast()` here would constrain its argument with a fuzzy result type (something like "must be compatible with an integer"), and that result type would have to be compatible with the actual type of `val`.  I think the source of my confusion is that a "result type" is not a different kind of "type", but is a way (via "results") of generating concrete types.

I guess its also a bit surprising to me that the result type can generally be precisely computed.  There aren't cases of overloading or ambiguity where the "result type" is a set of types?  (Especially with Zig's proliferation of `uN` and `iN` types.)  

Also, then the phrase in the comment "... permitted to be any integer type" isn't about the "result type"?  Its a different constraint?

Can you perhaps expand this example to be more specific about how `@intCast()` does (or doesn't) define a result type for its parameter?   (Functions normally can constrain their arguments, right?)

(Please feel free to ignore my comments if they're not resonating with you.  I don't mean to waste your time.)

Reply:
Don't worry, you're not wasting my time! Outside perspectives are valuable here; since I'm very intricately familiar with this system, it can be tricky for me to notice when things are unclear.

"permitted to be any integer type" is not referring to the result type - it's trying to explain why `@intCast` *can't* provide a result type to its parameter. In essence, because you could not write `@intCast` in Zig without an `anytype` parameter, we can't provide a result type to the argument here.

> There aren't cases of overloading or ambiguity where the "result type" is a set of types?

There definitely are, but RLS in its current form does not have a way to describe them. In general, they actually wouldn't be useful to represent - if we have a result type of "any integer", how could we actually use that? It doesn't help you with any specific coercions (aside from I suppose C pointer to `usize` but C pointers are silly) or casts.

The main place where it would be handy today is vectors - for instance, you can't currently write `my_vector * @splat(3)`, because we can't provide a specific result type for the RHS, even though we know the vector length which is *kind of* sufficient information. The thing is, every such case has to be explicitly codified into the specification of RLS, so it's not trivial. Before introducing something like that, I'd like to wait for integer types to settle down - specifically, I'd like to wait for a decision on #3806, since that drastically changes when and how we can provide result types when integers are involved.

Reply:
Ah, this raises another aspect of RLS: what is the impact/utility of the type inference?  I assumed it was for type checking (i.e., to give users a useful error message when they mistakenly pass a string to `@intCast()` or somesuch), but it sounds like it used to inform the code generation?